cpu-o3: delete pending wake events if cancel

happy-lx opened the pull request:
I encountered a bug with the following scenario:

```asm
fld fa4, 0(a0)
fmul fa5, fa5, fa4
fst fa5, 0(a0)
```

Bug scenario:
1. After `fld` is scheduled, it speculatively wakes up `fmul`.
2. After `fmul` is scheduled, a speculative wake-up event (`event1`) for `fst` is generated, and the actual wake-up occurs after a few cycles of delay.
3. `fld` is cancelled, which cancels the speculative wake-up for `fmul`, setting `fmul` as canceled, **but `event1` is not cleared.**
4. When `fld` is reissued, it completes and deterministically wakes up `fmul`, when `fmul` schedules, **`fmul->canceled()` is false**. At the same tick, `event1` starts to execute and wakes up `fst`.
5. When `fst` is scheduled, it is unable to obtain data from the bypass network(because the wake-up relationship between `fmul` and `fst` is broken).

The root cause of this bug is that **the speculative wake-up event was not canceled along with the instruction when it was canceled.**

This PR does modifications that when an instruction is canceled, its pending speculative wake-up events are also canceled.

Change-Id: Ib4c325f42de5874c4cc8c40c4389b9c8985df9f3

tastynoob commented on src/cpu/o3/issue_queue.cc:
```diff
@@ -606,15 +606,16 @@ IssueQue::doSquash(const InstSeqNum seqNum)
 Scheduler::SpecWakeupCompletion::process()
 {
     to_issue_queue->wakeUpDependents(inst, true);
+    scheduler->specWakeEvents.erase(this);
```
Could we save the iterator pointer to delere it faster

tastynoob commented on src/cpu/o3/issue_queue.cc:
```diff
@@ -983,6 +986,16 @@ Scheduler::loadCancel(const DynInstPtr& inst)
+            } else {
+                it++;
+            }
+        }
```
Same as above

tastynoob commented on src/cpu/o3/issue_queue.hh:
```diff
@@ -193,11 +194,12 @@ class Scheduler : public SimObject
     {
+      public:
         DynInstPtr inst;
+        Scheduler* scheduler;
```
Here only to save specWakeEvents's pointer

happy-lx committed:
cpu-o3: delete pending wake events if cancel

If this instruction is cancelled, the associated wake event should be descheduled.

Change-Id: I595541aa5f96163350aa5f6e3825f78520a0e660
